coderd/oauth2provider/revoke.go

```diff
@@ -1,44 +1,223 @@
+					slog.F("client_id", app.ID.String()),
+					slog.F("app_name", app.Name))
+				httpapi.WriteOAuth2Error(ctx, rw, http.StatusBadRequest, "invalid_request", "Invalid token format")
+				return
+			}
+			logger.Error(ctx, "token revocation failed with internal server error",
+				slog.Error(err),
+				slog.F("client_id", app.ID.String()),
+				slog.F("app_name", app.Name))
+			httpapi.WriteOAuth2Error(ctx, rw, http.StatusInternalServerError, "server_error", "Internal server error")
 			return
 		}
-		rw.WriteHeader(http.StatusNoContent)
+
+		// RFC 7009: successful revocation returns HTTP 200
+		rw.WriteHeader(http.StatusOK)
+	}
+}
+
+func revokeRefreshTokenInTx(ctx context.Context, db database.Store, token string, appID uuid.UUID) error {
+	// Parse the refresh token using the existing function
+	parsedToken, err := parseFormattedSecret(token)
+	if err != nil {
+		return ErrInvalidTokenFormat
 	}
+
+	// Try to find refresh token by prefix
+	//nolint:gocritic // Using AsSystemOAuth2 for OAuth2 public token revocation endpoint
+	dbToken, err := db.GetOAuth2ProviderAppTokenByPrefix(dbauthz.AsSystemOAuth2(ctx), []byte(parsedToken.prefix))
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// Token not found - return success per RFC 7009 (don't reveal token existence)
+			return nil
+		}
+		return xerrors.Errorf("get oauth2 provider app token by prefix: %w", err)
+	}
+
+	// Verify ownership
+	//nolint:gocritic // Using AsSystemOAuth2 for OAuth2 public token revocation endpoint
+	appSecret, err := db.GetOAuth2ProviderAppSecretByID(dbauthz.AsSystemOAuth2(ctx), dbToken.AppSecretID)
+	if err != nil {
+		return xerrors.Errorf("get oauth2 provider app secret: %w", err)
+	}
+	if appSecret.AppID != appID {
+		return ErrTokenNotBelongsToClient
+	}
+
+	// Delete the associated API key, which should cascade to remove the refresh token
+	// According to RFC 7009, when a refresh token is revoked, associated access tokens should be invalidated
+	//nolint:gocritic // Using AsSystemOAuth2 for OAuth2 public token revocation endpoint
+	err = db.DeleteAPIKeyByID(dbauthz.AsSystemOAuth2(ctx), dbToken.APIKeyID)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return xerrors.Errorf("delete api key: %w", err)
+	}
+
+	return nil
+}
+
+// parsedAPIKey represents the components of an API key token
+type parsedAPIKey struct {
+	keyID  string // The API key ID for database lookup
+	secret string // The secret part for verification
+}
+
+// parseAPIKeyToken parses an API key token following the encoder/decoder pattern
+func parseAPIKeyToken(token string) (parsedAPIKey, error) {
+	parts := strings.SplitN(token, "-", 2)
+	if len(parts) != 2 {
+		return parsedAPIKey{}, xerrors.Errorf("incorrect number of parts: %d", len(parts))
+	}
+	if parts[0] == "" || parts[1] == "" {
+		return parsedAPIKey{}, xerrors.New("empty key ID or secret")
+	}
+	return parsedAPIKey{
+		keyID:  parts[0],
+		secret: parts[1],
+	}, nil
+}
```

Comment:
We should not reimplement https://github.com/coder/coder/blob/f3d950d9177846ff1fdfc4b070524a642f8c737a/coderd/httpmw/apikey.go#L746-L767

(I am fixing)